src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -131,11 +131,25 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
                 var corLibrary = this.ContainingAssembly.CorLibrary;
                 var conversions = new TypeConversions(corLibrary);
                 var location = singleDeclaration.NameLocation;
+                var seenInterfaces = PooledHashSet<NamedTypeSymbol>.GetInstance();
 
                 foreach (var @interface in interfaces)
                 {
                     @interface.CheckAllConstraints(conversions, location, diagnostics);
+
+                    foreach (var seenInterface in seenInterfaces)
+                    {
+                        if (@interface.Equals(seenInterface, TypeCompareKind.IgnoreTupleNames))
```

Comment:
Consider using the comparer as the comparer for the `HashSet` (or `Dictionary` if necessary) so we don't have to loop over the collection to find the earlier interface. If using a custom comparer means we can't use a pooled collection, perhaps allocation the collection only if there is more than one interface.